File: ArxivExpress/ArxivExpress/Features/SearchArticles/Data/SearchQuery.cs
// ****************************************************************************
//    File "SearchQuery.cs"
//    Copyright © Dmitry Morozov 2022
//    If you want to use this file please contact me by dvmorozov@hotmail.com.
// ****************************************************************************

namespace ArxivExpress.Features.SearchArticles
{
    //  TODO: Convert SearchQuery to singleton.
    public class SearchQuery
    {
        public string SearchTerm;
        public string Prefix;
        public string ResultsPerPage;

        //  Mutually exclusive.
        public bool SortByRelevance;
        public bool SortByLastUpdatedDate;
        public bool SortBySubmittedDate;

        //  Mutually exclusive.
        public bool SortOrderAscending;
        public bool SortOrderDescending;

        public uint PageNumber;

        public SearchQuery(string searchTerm = "", string prefix = "all")
        {
            FillDefaultValues(searchTerm, prefix);
        }

        private void FillDefaultValues(string searchTerm, string prefix)
        {
            SearchTerm = searchTerm;
            ResultsPerPage = "25";
            Prefix = prefix;

            SortByRelevance = false;
            SortByLastUpdatedDate = false;
            SortBySubmittedDate = true;

            SortOrderAscending = false;
            SortOrderDescending = true;

            PageNumber = 0;
        }

        private string GetQuerySortBy()
        {
            if (SortByRelevance)
            {
                return "relevance";
            }
            else if (SortByLastUpdatedDate)
            {
                return "lastUpdatedDate";
            }
            else //if (SortBySubmittedDate)
            {
                return "submittedDate";
            }
        }

        private string GetQuerySortOrder()
        {
            if (SortOrderAscending)
            {
                return "ascending";
            }
            else //if (SortOrderDescending)
            {
                return "descending";
            }
        }

        public uint GetResultsPerPage()
        {
            if (!uint.TryParse(ResultsPerPage, out uint resultsPerPage))
            {
                resultsPerPage = 25;
            }

            return resultsPerPage;
        }

        public string GetQueryString()
        {
            var queryString = "http://export.arxiv.org/api/query?search_query=";

            if (Prefix != null && Prefix != string.Empty)
            {
                queryString += Prefix;
            }
            else
            {
                queryString += "all";
            }

            if (SearchTerm != null && SearchTerm != string.Empty)
            {
                queryString += ":\"" + SearchTerm + "\"";
            }

            var resultsPerPage = GetResultsPerPage();

            queryString += "&start=" + (PageNumber * resultsPerPage).ToString();
            queryString += "&max_results=" + resultsPerPage.ToString();
            queryString += "&sortBy=" + GetQuerySortBy();
            queryString += "&sortOrder=" + GetQuerySortOrder();

            return queryString;
        }
    }
}

File: ArxivExpressFlutter/lib/features/search_articles/data/search_query.dart
// ****************************************************************************
//    File "search_query.dart"
//    Copyright © Dmitry Morozov 2025
//    Flutter reimplementation of ArxivExpress Xamarin app
// ****************************************************************************

class SearchQuery {
  String searchTerm;
  String prefix;
  String resultsPerPage = "25";

  // Mutually exclusive
  bool sortByRelevance = false;
  bool sortByLastUpdatedDate = false;
  bool sortBySubmittedDate = true;

  // Mutually exclusive
  bool sortOrderAscending = false;
  bool sortOrderDescending = true;

  int pageNumber = 0;

  SearchQuery({
    this.searchTerm = "",
    this.prefix = "all",
  }) {
    _fillDefaultValues();
  }

  void _fillDefaultValues() {
    resultsPerPage = "25";

    sortByRelevance = false;
    sortByLastUpdatedDate = false;
    sortBySubmittedDate = true;

    sortOrderAscending = false;
    sortOrderDescending = true;

    pageNumber = 0;
  }

  String _getQuerySortBy() {
    if (sortByRelevance) {
      return "relevance";
    } else if (sortByLastUpdatedDate) {
      return "lastUpdatedDate";
    } else {
      return "submittedDate";
    }
  }

  String _getQuerySortOrder() {
    if (sortOrderAscending) {
      return "ascending";
    } else {
      return "descending";
    }
  }

  int getResultsPerPage() {
    final parsed = int.tryParse(resultsPerPage);
    return parsed ?? 25;
  }

  String getQueryString() {
    var queryString = "http://export.arxiv.org/api/query?search_query=";

    if (prefix.isNotEmpty) {
      queryString += prefix;
    } else {
      queryString += "all";
    }

    if (searchTerm.isNotEmpty) {
      queryString += ":$searchTerm";
    }

    queryString += "&start=$pageNumber";
    queryString += "&max_results=${getResultsPerPage()}";
    queryString += "&sortBy=${_getQuerySortBy()}";
    queryString += "&sortOrder=${_getQuerySortOrder()}";

    return queryString;
  }
}